### PR TITLE
Factor Linux specific functions out of zpool command

### PR DESCRIPTION
**Pull request by mattmacy**
Signed-of-by: Matt Macy <mmacy@FreeBSD.org>

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: cmd/zpool/zpool_vdev.c [hunk 1]
```diff
@@ -829,7 +498,7 @@ get_replication(nvlist_t *nvroot, boolean_t fatal)
 			rep.zprl_children = 1;
 			rep.zprl_parity = 0;
 		} else {
-			uint64_t vdev_size;
+			int64_t vdev_size;
```
Comment (loli10K) [inline cmd/zpool/zpool_vdev.c]: Could you explain the reason behind all these "unit64_t" changes? Doesn't seem Linux specific functionality.

Comment (freqlabs) [reply cmd/zpool/zpool_vdev.c]: `labs(size - vdev_size)` triggers a clang warning "taking the absolute value of unsigned type 'unsigned long' has no effect"

Comment (allanjude) [reply cmd/zpool/zpool_vdev.c]: > 
> 
> `labs(size - vdev_size)` triggers a clang warning "taking the absolute value of unsigned type 'unsigned long' has no effect"

Then the right solution might be to drop the labs() rather than change the type. The type should definitely stay uint64_t

Comment (mattmacy) [reply cmd/zpool/zpool_vdev.c]: > > `labs(size - vdev_size)` triggers a clang warning "taking the absolute value of unsigned type 'unsigned long' has no effect"
> 
> Then the right solution might be to drop the labs() rather than change the type. The type should definitely stay uint64_t

Not if size can be less than vdev_size.

Comment (mattmacy) [reply cmd/zpool/zpool_vdev.c]: `size` and `vdev_size` will always be representable by 63-bits so keeping it unsigned is gratuitous if we aim to take the difference of the two.

Comment (freqlabs) [reply cmd/zpool/zpool_vdev.c]: The difference between the sizes could be positive or negative. Leaving the types as unsigned means the result overflows when the difference is negative and removing the labs() means we'll have introduced a bug. The subtraction results in the correct value when the unsigned integer is interpreted as a signed integer by labs().

Clang doesn't see that we're doing a subtraction and abusing the types. It sees the result of the subtraction, an unsigned value, being passed to an absolute value function and emits a warning which we treat as an error.

Comment (behlendorf) [reply cmd/zpool/zpool_vdev.c]: We've ended up making similar fixes throughout the code for issues like this; e.g. when it's expected that the value can may be negative.  But since this change is independent of the proposed refactoring, let's open a new PR with only this fix and a brief explanation.  @freqlabs's comment above would be great.

Comment (freqlabs) [reply cmd/zpool/zpool_vdev.c]: #9335

File: cmd/zpool/os/linux/zpool_vdev_os.c [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+	{"ATA     WDC WD15EARS-00Z", 4096},
+	{"ATA     WDC WD20EARS-00M", 4096},
+	{"ATA     WDC WD20EARS-00S", 4096},
+	{"ATA     WDC WD20EARS-00Z", 4096},
+	{"ATA     WDC WD1600BEVT-0", 4096},
+	{"ATA     WDC WD2500BEVT-0", 4096},
+	{"ATA     WDC WD3200BEVT-0", 4096},
+	{"ATA     WDC WD5000BEVT-0", 4096},
+	/* Virtual disks: Assume zvols with default volblocksize */
+#if 0
```
Comment (kithrup) [inline cmd/zpool/os/linux/zpool_vdev_os.c]: Why is this if'd out but present?

Comment (mattmacy) [reply cmd/zpool/os/linux/zpool_vdev_os.c]: mismerge - thanks

Comment (freqlabs) [reply cmd/zpool/os/linux/zpool_vdev_os.c]: It moved here from zpool_vdev.c:217

Comment (behlendorf) [reply cmd/zpool/os/linux/zpool_vdev_os.c]: Let's simply remove these dead entries.

Is there any benefit to keeping this list of known problematic devices in the common code?  Or does FreeBSD already keep a similar, hopefully more extensive list, lower down in the block device stack?

File: cmd/zpool/Makefile.am [hunk 3]
```diff
@@ -2,7 +2,8 @@ include $(top_srcdir)/config/Rules.am
 
 DEFAULT_INCLUDES += \
 	-I$(top_srcdir)/include \
-	-I$(top_srcdir)/lib/libspl/include
+	-I$(top_srcdir)/lib/libspl/include \
+	-I.
```
Comment (behlendorf) [inline cmd/zpool/Makefile.am]: Why was this needed?  `zpool_util.h` hasn't moved.

**Comment by behlendorf** [comment]
Could you rebase this on master now that #9335 was merged.

File: cmd/zpool/Makefile.am (hunk 3 above)
Comment (mattmacy) [reply cmd/zpool/Makefile.am]: It's relative location has changed.

**Comment by mattmacy** [comment]
> Could you rebase this on master now that #9335 was merged.

Done

File: cmd/zpool/zpool_util.h [hunk 4]
```diff
@@ -125,6 +125,12 @@ vdev_cmd_data_list_t *all_pools_for_each_vdev_run(int argc, char **argv,
 
 void free_vdev_cmd_data_list(vdev_cmd_data_list_t *vcdl);
 
+int check_device(const char *path, boolean_t force,
+    boolean_t isspare, boolean_t iswholedisk);
+boolean_t check_sector_size_database(char *path, int *sector_size);
+void vdev_error(const char *fmt, ...);
+int check_file(const char *file, boolean_t force, boolean_t isspare);
```
Comment (behlendorf) [inline cmd/zpool/zpool_util.h]: After taking another look at this, I think a nicer approach would be to move this functionality in to `libzutil`.  Adding `zfs_isnumber()` to `lib/libzutil/zutil_nicenum.c` would make sense and be generally useful.  Plus, `lib/libzutil/zutil_import.c` alreadys contain platform specific code and would be a natural home for `check_device()`, `check_file()`, `check_sector_size_database()` after the refactoring in #9356 is settled.

This would have the advantage of keeping the `zpool` command itself largely portable with the platform specific being contained in the library.

**Comment by freqlabs** [comment]
Rebased, moved isnumber to libzutil as zfs_isnumber.

File: cmd/zpool/zpool_util.h (hunk 4 above)
Comment (freqlabs) [reply cmd/zpool/zpool_util.h]: The vdev checks were difficult to move to libzutil when I tried. It introduces a circular dependency with libzfs. I've left them where they are for now.

Comment (behlendorf) [reply cmd/zpool/zpool_util.h]: OK, thanks for investigating.  Then this LGTM.
